open-feature/rust-sdk
# feat: Add hooks implementation

## This PR
This pull request introduces the implementation of Hooks in the OpenFeature SDK following the OpenFeature standard. The changes include the addition of hooks to various stages in the feature flag evaluation process, such as before, after, error, and finally. The implementation provides capabilities to add hooks at the global, client, provider, and invocation levels, which allows for extensive customizability and control over the feature flag lifecycle.

- Adds hooks to support the OpenFeature specification for flag evaluation.
- Implements the capability to register hooks at various levels including global, client, provider, and invocation.
- Introduces logging for lifecycle events using hooks.
- Provides test cases for the new hooks capability to ensure compliance with the OpenFeature specification.

### Related Issues
Fixes #5 

### Notes
- Make sure logging is appropriately configured in the consuming application to take advantage of the logging hooks.
- This implementation supports extensibility for additional hooks in the future by following the standardized hook interface.

### Follow-up Tasks
- Consider implementing example hooks that demonstrate different use cases, like metrics collection or error reporting.
- Evaluate if default hooks should be provided in the SDK for common use cases.

### How to test
1. Build and run the example included in `examples/hooks.rs` to see a working implementation of the hooks in action.
2. Review test cases provided in `src/hooks/mod.rs` to ensure they cover expected behaviors and edge cases.
3. Ensure that the hooks are called in the correct order and can handle errors gracefully.
